src/api/networking.rs

```diff
@@ -854,3 +1023,615 @@ fn socket_address(
+//% * 1 on error   - The error ID is written to **id_u64_ptr**
+//%
+//% Creates a new UDP socket, which will be bound to the specified address. The returned socket
+//% is ready for receiving messages.
+//%
+//% Binding with a port number of 0 will request that the OS assigns a port to this socket. The
+//% port allocated can be queried via the `udp_local_addr` method.
+//%
+//% Traps:
+//% * If **addr_type** is neither 4 or 6.
+//% * If **addr_u8_ptr** is outside the memory
+//% * If **id_u64_ptr** is outside the memory.
+fn udp_bind(
+    mut caller: Caller<ProcessState>,
+    addr_type: u32,
+    addr_u8_ptr: u32,
+    port: u32,
+    flow_info: u32,
+    scope_id: u32,
+    id_u64_ptr: u32,
+) -> Box<dyn Future<Output = Result<u32, Trap>> + Send + '_> {
+    Box::new(async move {
+        let memory = get_memory(&mut caller)?;
+        let socket_addr = socket_address(
+            &caller,
+            &memory,
+            addr_type,
+            addr_u8_ptr,
+            port,
+            flow_info,
+            scope_id,
+        )?;
+        let (udp_listener_or_error_id, result) = match UdpSocket::bind(socket_addr).await {
+            Ok(listener) => (
+                caller
+                    .data_mut()
+                    .resources
+                    .udp_sockets
+                    .add(Arc::new(listener)),
+                0,
+            ),
+            Err(error) => (caller.data_mut().errors.add(error.into()), 1),
+        };
+        memory
+            .write(
+                &mut caller,
+                id_u64_ptr as usize,
+                &udp_listener_or_error_id.to_le_bytes(),
+            )
+            .or_trap("lunatic::networking::udp_bind")?;
+
+        Ok(result)
+    })
+}
+
+//% lunatic::networking::drop_udp_listener(tcp_listener_id: i64)
+//%
+//% Drops the UCP listener resource.
+//%
+//% Traps:
+//% * If the UDP listener ID doesn't exist.
+fn drop_udp_socket(mut caller: Caller<ProcessState>, udpp_listener_id: u64) -> Result<(), Trap> {
+    caller
+        .data_mut()
+        .resources
+        .udp_sockets
+        .remove(udpp_listener_id)
+        .or_trap("lunatic::networking::drop_udp_socket")?;
+    Ok(())
+}
+
+//% lunatic::networking::udp_receive(
+//%     socket_id: u64,
+//%     buffer_ptr: u32,
+//%     buffer_len: u32,
+//%     timeout: u32,
+//%     i64_opaque_ptr: u32,
+//% ) -> i32
+//%
+//% Returns:
+//% * 0 on success    - The number of bytes read is written to **opaque_ptr**
+//% * 1 on error      - The error ID is written to **opaque_ptr**
+//% * 9027 on timeout - The socket receive timed out.
+//%
+//% Reads data from the connected udp socket and writes it to the given buffer.
+//%
+//% Traps:
+//% * If the stream ID doesn't exist.
+//% * If **buffer_ptr + buffer_len** is outside the memory.
+//% * If **i64_opaque_ptr** is outside the memory.
+//% * If the socket hasn't been connected yet. (use socket.connect() to connect to an address)
```

Comment:
I don't think this function will trap if the socket is not connected yet. It will just just return an error. On line `1158` it captures the error and writes it the the return value.